libenv/sysinfo.c

```diff
@@ -2537,24 +2535,42 @@ static void Xen_Cpuid(uint32_t idx, uint32_t *eax, uint32_t *ebx, uint32_t *ecx,
 
 /******************************************************************/
 
-static int Xen_Hv_Check(void)
+static bool Xen_Hv_Check(void)
 {
-    uint32_t eax;
+    /* CPUID interface to Xen from arch-x86/cpuid.h:
+     * Leaf 1 (0x40000000)
+     * EAX: Largest Xen-information leaf. All leaves up to an including @EAX
+     *   are supported by the Xen host.
+     * EBX-EDX: "XenVMMXenVMM" signature, allowing positive identification
+     *   of a Xen host.
+     *
+     * Additional information can be found in the Hypervisor CPUID
+     * Interface Proposal (https://lkml.org/lkml/2008/10/1/246)
+     */
+
+    uint32_t eax, base;
     union
     {
         uint32_t u[3];
         char s[13];
     } sig = {{0}};
 
-    Xen_Cpuid(0x40000000, &eax, &sig.u[0], &sig.u[1], &sig.u[2]);
-
-    if (strcmp("XenVMMXenVMM", sig.s) || (eax < 0x40000002))
+    for (base = 0x40000000; base < 0x40010000; base += 0x100)
```

Comment:
To properly detect Xen you need to scan the leaves between 0x40000000 to 0x40010000. Take a look at [xen-detect.c](https://github.com/mirage/xen/blob/master/tools/misc/xen-detect.c#L61) for some additional details.

I am not quite sure why this is segfaulting and since I don't have access to a box running RHEL 4, per your suggestion on #cfengine-dev, the new pull request will just return false if the redhat_4 or centos_4 class is defined.

A quick test:

```
# /var/cfengine/bin/cf-promises --show-classes --debug | grep -i Xen
 verbose: Found Xen CPUID Leaf. eax=40000002 at base 40000000
 verbose: This appears to be a xen hv system.
   debug: Setting hard class: default:xen
   debug: Setting hard class: default:xen_domu_hv
 verbose: C: discovered hard class xen
 verbose: C: discovered hard class xen_domu_hv
xen                                                          inventory,attribute_name=Virtual host,source=agent,hardclass
xen_domu_hv                                                  source=agent,hardclass

# /var/cfengine/bin/cf-promises --show-classes -Dredhat_4 --debug | grep -i Xen
   debug: Skipping Xen_Hv_Check() to avoid a segfault on RHEL 4
```